### PR TITLE
Srw meta

### PR DESCRIPTION
**Pull request by antmat**
Ability to pass headers from client code.
Seems like it's done now

### CODE REVIEW

File: include/cocaine/rpc/queue.hpp [hunk 1]
```diff
@@ -68,7 +68,7 @@ class message_queue {
 public:
     template<class Event, class... Args>
     void
-    append(Args&&... args) {
+    append(hpack::header_storage_t headers, Args&&... args) {
```
Comment (3Hren) [inline include/cocaine/rpc/queue.hpp]: No, we shouldn't break the API. Add overloads instead.

File: include/cocaine/rpc/slot/deferred.hpp [hunk 2]
```diff
@@ -113,21 +113,21 @@ struct deferred {
         std::is_constructible<T, Args...>::value,
         deferred&
     >::type
-    write(Args&&... args) {
-        outbox->synchronize()->template append<typename protocol::value>(std::forward<Args>(args)...);
+    write(hpack::header_storage_t headers, Args&&... args) {
```
Comment (3Hren) [inline include/cocaine/rpc/slot/deferred.hpp]: The same.

File: include/cocaine/rpc/slot/deferred.hpp [hunk 3]
```diff
@@ -113,21 +113,21 @@ struct deferred {
...
+        outbox->synchronize()->template append<typename protocol::value>(std::move(headers), std::forward<Args>(args)...);
         return *this;
     }
 
     deferred&
-    abort(const std::error_code& ec, const std::string& reason) {
-        outbox->synchronize()->template append<typename protocol::error>(ec, reason);
+    abort(hpack::header_storage_t headers, const std::error_code& ec, const std::string& reason) {
```
Comment (3Hren) [inline include/cocaine/rpc/slot/deferred.hpp]: And here.

File: include/cocaine/rpc/slot/deferred.hpp [hunk 4]
```diff
@@ -113,21 +113,21 @@ struct deferred {
...
+        outbox->synchronize()->template append<typename protocol::error>(std::move(headers), ec, reason);
         return *this;
     }
 
 #if defined(__clang__)
     deferred&
-    abort(const std::error_code& ec) {
-        outbox->synchronize()->template append<typename protocol::error>(ec);
+    abort(hpack::header_storage_t headers, const std::error_code& ec) {
```
Comment (3Hren) [inline include/cocaine/rpc/slot/deferred.hpp]: Here.

File: include/cocaine/rpc/slot/deferred.hpp [hunk 5]
```diff
@@ -156,22 +156,22 @@ struct deferred<void> {
     { }
 
     deferred&
-    abort(const std::error_code& ec, const std::string& reason) {
-        outbox->synchronize()->append<protocol::error>(ec, reason);
+    abort(hpack::header_storage_t headers, const std::error_code& ec, const std::string& reason) {
```
Comment (3Hren) [inline include/cocaine/rpc/slot/deferred.hpp]: Here.

File: include/cocaine/rpc/slot/deferred.hpp [hunk 6]
```diff
@@ -156,22 +156,22 @@ struct deferred<void> {
...
+        outbox->synchronize()->append<protocol::error>(std::move(headers), ec, reason);
         return *this;
     }
 
 #if defined(__clang__)
     deferred&
-    abort(const std::error_code& ec) {
-        outbox->synchronize()->append<protocol::error>(ec);
+    abort(hpack::header_storage_t headers, const std::error_code& ec) {
```
Comment (3Hren) [inline include/cocaine/rpc/slot/deferred.hpp]: And even here.

File: include/cocaine/rpc/slot/deferred.hpp [hunk 7]
```diff
@@ -156,22 +156,22 @@ struct deferred<void> {
...
+        outbox->synchronize()->append<protocol::error>(std::move(headers), ec);
         return *this;
     }
 #endif
 
     deferred&
-    close() {
-        outbox->synchronize()->append<protocol::value>();
+    close(hpack::header_storage_t headers) {
```
Comment (3Hren) [inline include/cocaine/rpc/slot/deferred.hpp]: Finally here.

File: include/cocaine/rpc/slot/streamed.hpp [hunk 8]
```diff
@@ -43,28 +43,28 @@ struct streamed {
         std::is_constructible<T, Args...>::value,
         streamed&
     >::type
-    write(Args&&... args) {
-        outbox->synchronize()->template append<typename protocol::chunk>(std::forward<Args>(args)...);
+    write(hpack::header_storage_t headers, Args&&... args) {
```
Comment (3Hren) [inline include/cocaine/rpc/slot/streamed.hpp]: Well, here too.

File: include/cocaine/rpc/slot/streamed.hpp [hunk 9]
```diff
@@ -43,28 +43,28 @@ struct streamed {
...
+        outbox->synchronize()->template append<typename protocol::chunk>(std::move(headers), std::forward<Args>(args)...);
         return *this;
     }
 
     streamed&
-    abort(const std::error_code& ec, const std::string& reason) {
-        outbox->synchronize()->template append<typename protocol::error>(ec, reason);
+    abort(hpack::header_storage_t headers, const std::error_code& ec, const std::string& reason) {
```
Comment (3Hren) [inline include/cocaine/rpc/slot/streamed.hpp]: And so on.

File: include/cocaine/rpc/slot/streamed.hpp [hunk 10]
```diff
@@ -43,28 +43,28 @@ struct streamed {
...
+        outbox->synchronize()->template append<typename protocol::error>(std::move(headers), ec, reason);
         return *this;
     }
 
 #if defined(__clang__)
     streamed&
-    abort(const std::error_code& ec) {
-        outbox->synchronize()->template append<typename protocol::error>(ec);
+    abort(hpack::header_storage_t headers, const std::error_code& ec) {
```
Comment (3Hren) [inline include/cocaine/rpc/slot/streamed.hpp]: Why we ever need ifdef for clang BTW?

File: include/cocaine/rpc/slot/streamed.hpp [hunk 11]
```diff
@@ -43,28 +43,28 @@ struct streamed {
...
+        outbox->synchronize()->template append<typename protocol::error>(std::move(headers), ec);
         return *this;
     }
 #endif
 
     streamed&
-    close() {
-        outbox->synchronize()->template append<typename protocol::choke>();
+    close(hpack::header_storage_t headers) {
```
Comment (3Hren) [inline include/cocaine/rpc/slot/streamed.hpp]: But here too.

File: include/cocaine/rpc/upstream.hpp [hunk 12]
```diff
@@ -45,12 +45,23 @@ class basic_upstream_t {
     void
     send(Args&&... args);
 
+    template<class Event, class... Args>
+    void
+    send(hpack::header_storage_t headers, Args&&... args);
```
Comment (3Hren) [inline include/cocaine/rpc/upstream.hpp]: Definetely here.

File: include/cocaine/rpc/upstream.hpp [hunk 13]
```diff
@@ -45,12 +45,23 @@ class basic_upstream_t {
...
+
     /* none_t if upstream belongs to server side */
     boost::optional<trace_t> client_trace;
 };
 
 template<class Event, class... Args>
 void
+basic_upstream_t::send(hpack::header_storage_t headers, Args&&... args) {
```
Comment (3Hren) [inline include/cocaine/rpc/upstream.hpp]: Yep, here.

File: src/header.cpp [hunk 14]
```diff
@@ -190,6 +190,17 @@ header_t::http2_size() const {
...
 
+const std::vector<header_t>&
+header_storage_t::get_headers() const {
+    return headers;
+}
+
+void
+header_storage_t::push_back(const header_t& header) {
+    headers.push_back(header);
+    zone.rebind_header(headers.back());
```
Comment (3Hren) [inline src/header.cpp]: Looks strange. What does this line do?

Comment (antmat) [inline src/header.cpp]: copies header data to zone, and points header to that data.

File: include/cocaine/rpc/upstream.hpp (hunk 12 above)
Comment (antmat) [inline include/cocaine/rpc/upstream.hpp]: nope, it's already overload

File: include/cocaine/rpc/upstream.hpp (hunk 13 above)
Comment (antmat) [inline include/cocaine/rpc/upstream.hpp]: and here it's already overload

File: include/cocaine/rpc/slot/streamed.hpp (hunk 11 above)
Comment (antmat) [inline include/cocaine/rpc/slot/streamed.hpp]: In fact there will be plenty of overloads.
What do you think about adding set_headers method to deferred/streamed?

Comment (3Hren) [inline include/cocaine/rpc/slot/streamed.hpp]: [Setters are antipattern](http://www.yegor256.com/2014/09/16/getters-and-setters-are-evil.html), I strongly disrespect them.

**Comment by 3Hren** [comment]
LGTM

File: include/cocaine/rpc/slot/deferred.hpp [hunk 15]
```diff
@@ -118,19 +118,27 @@ struct deferred {
         return *this;
     }
 
+    template<class... Args>
+    typename std::enable_if<
+    std::is_constructible<T, Args...>::value,
+    deferred&
```
Comment (3Hren) [inline include/cocaine/rpc/slot/deferred.hpp]: Indent.

File: include/cocaine/rpc/upstream.hpp [hunk 16]
```diff
@@ -82,6 +93,20 @@ class upstream {
 
     template<class Event, class... Args>
     upstream<typename io::event_traits<Event>::dispatch_type>
+    send(hpack::header_storage_t headers, Args&&... args) {
+        static_assert(
+        std::is_same<typename Event::tag, Tag>::value,
```
Comment (3Hren) [inline include/cocaine/rpc/upstream.hpp]: Indent!

**Pull request merged**